[Merged by Bors] - feat(Topology/UniformSpace): add `UniformContinuous.iterate `

pitmonticone opened the pull request:
This PR adds the theorem `UniformContinuous.iterate` claiming that if a function `T` is uniformly continuous in a uniform space `β`, then its `n`-th iterate `T^[n]` is also uniformly continuous.

Co-authored-by: @D-Thomine 

---

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

sgouezel commented on Mathlib/Topology/UniformSpace/Basic.lean:
```diff
@@ -1120,6 +1120,15 @@ nonrec theorem UniformContinuous.comp [UniformSpace β] [UniformSpace γ] {g : 
 
+/--If a function `T` is uniformly continuous in a uniform space `β`,
+then its `n`-th iterate `T^[n]` is also uniformly continuous.-/
+theorem uniformContinuous_ite [UniformSpace β] (T : β → β) (n : ℕ) (h : UniformContinuous T) :
```
The name is bad, because `ite` means `if then else`. What about `UniformContinuous.iterate`?

pitmonticone replied:
Ok.

sgouezel commented on Mathlib/Topology/UniformSpace/Basic.lean:
```diff
@@ -1120,6 +1120,15 @@ nonrec theorem UniformContinuous.comp [UniformSpace β] [UniformSpace γ] {g : 
+then its `n`-th iterate `T^[n]` is also uniformly continuous.-/
+theorem uniformContinuous_ite [UniformSpace β] (T : β → β) (n : ℕ) (h : UniformContinuous T) :
+    UniformContinuous T^[n] := by
+  induction' n with n hn
```
Does `induction` work instead of `induction'`? (It is more canonical).

pitmonticone replied:
No, it doesn't work if I naively substitute it for `induction'`.

sgouezel replied:
Sure, naive substitution won't work, but `induction'` is suboptimal and will probably be removed some day, so better learn the new syntax. Can you figure it out (for instance looking for `induction n` around)? If you can't, I can show you how it should be done, but I think it's more instructive to let you find out.

pitmonticone replied:
Sure, I'll do it as soon as I can.

pitmonticone replied:
Done.

sgouezel commented on Mathlib/Topology/UniformSpace/Basic.lean:
```diff
@@ -1120,6 +1120,15 @@ nonrec theorem UniformContinuous.comp [UniformSpace β] [UniformSpace γ] {g : 
+  induction' n with n hn
+  · exact uniformContinuous_id
+  · exact Function.iterate_succ _ _ ▸ UniformContinuous.comp hn h
+#align uniform_continuous_ite uniformContinuous_ite
```
No #align for new statements, just remove this line.

pitmonticone replied:
Done.

pitmonticone commented:
Ok, I should have addressed all the review requests (apart from the `induction` vs. `induction'` that seems to be minor). 

Thank you very much @sgouezel for your review!

pitmonticone commented:
Ok, I should have addressed all the review requests.

Thank you very much @sgouezel for your review!

sgouezel commented:
bors r+
Thanks!
